### ISSUE TITLE
[feature request] Better handling of file extensions

### ISSUE THREAD

**Issue by loganfsmyth** [babel/babel#8652]
## Feature Request

A request we've gotten in several different contexts is for Babel to have better file extension handling. This is particularly clear in the context of Typescript, because we don't want to process`.ts` files unless users have enabled `@babel/preset-typescript`. Unfortunately, Babel currently pretty much just processes whatever it is given and assumes that it will be able to parse it, so if it is given Typescript without the plugin enabled, it will just throw errors about unknown syntax.

Some examples:
* #3741
* #6027
* #7620

Essentially, these all come down to a primary question, which is, how does Babel decide what files to process? Before the release of 7.x, I experimented quite a bit around file extension handling. The behavior I lean toward is, Babel should be passed all files, and if it doesn't care about a file, it can behave like the file was `ignore`ed.

The primary concern from my point of view is, which stopped me from landing my PR in 7.x, was what about files have have non-standard extensions? For instance, `.vue` templates still get passed to Babel. Do we require all users of `vue` to enable a Babel plugin to opt into compilation of `.vue` files? The downside there is that `.vue` files in Webpack are only _sometimes_ JavaScript. Sometimes, it might be the actual Vue template file. Ideally Webpack would have some way to split up files, so we'd at least get a `.js.vue` file or something, but it's not really obvious what the right behavior is, and we were very close to releasing 7.x by the time I started thinking about this issue.

A few other questions:
* What should Babel do if no filename is given? (This may have answered itself, since we now require a filename to be given, in order to use `preset-typescript`, unless you pass an option).
* What about cases like `@babel/register` where we need to know the extensions up front in order to register the require hooks? We'd need a concept of a project root config, which we kind of have now with `configFile` and `babel.config.js`, but we have no infrastructure to query that project-wide config for things like registered file extensions. It also means that if you loaded `preset-typescript` in a `.babelrc`, we'd still always ignore `.ts` files, which isn't the best.
* Do we want extensions to have metadata associated with them? In 7.x for instance we hard-coded `.mjs` to be `sourceType: "module"` always, is that what we want in the long run, or should it be a configurable flag attached to each extension?
* Do we want this to try to handle _output_ extensions, for things like `gulp-babel` and `@babel/cli` that frequently write files back to disk after transformation?

It's definitely something we should think through for 8.x though.

**Comment by wyqydsyq** [comment]
I have a client/server project where I don't want to compile my server but I also want to get support for babel and TypeScript stuff, so using the CLI with `--extensions` is not an option for me.

I have an `index.js` entry point in my server which initialises `@babel/register` then loads the main TS file to kickstart the server.

The problem is that because Babel does not recognise `.ts` files and has no configurable option to enable that, all of my TypeScript files I'm using need to be saved as `.js` for `@babel/register` to pick them up.

`extensions` needs to be available as an option configurable via `.babelrc` and/or the object passed to `@babel/register`, this would enable my use case with the following as my `index.js` entry point instead of having to give all my TS files `.js` extensions:

```js
// ./server/index.js
require('@babel/register')({
  only: /\/server/,
  extensions: ['.ts'],
});
require('./server');
```

**Comment by loganfsmyth** [comment]
The snippet of code you posted is valid already. `extensions` is a valid option for `@babel/register`.

**Comment by wyqydsyq** [comment]
Right you are, I actually just pseudo-coded that without running it after trying to specify the `exensions` option in `.babelrc` without any luck, assuming it would behave the same via `@babel/register`.

In that case it seems support for configuring `extensions` via `.babelrc` should be implemented purely for consistency since the options is available via both CLI and `@babel/register`?

**Comment by loganfsmyth** [comment]
Yep, I'm sure we'll get there eventually, but it is potentially complicated to get right given our current design.

**Comment by sibelius** [comment]
are you gonna accept a PR that enable `extensions` on .babelrc

**Comment by loganfsmyth** [comment]
@sibelius We could potentially handle it in a `babel.config.js`, but I don't think we can realistically support it in `.babelrc` files because utilities like `@babel/register` need to know the list of extensions before Babel even starts compiling any files, and `.babelrc` file loading is entirely dependent on the path of a given file being compiled.

**Comment by deser** [comment]
So what guys? :)

**Comment by dandv** [comment]
Came here from #3741, which was locked, so I can't ask there: is it possible to configure the extensions via the config file?

**Comment by valoricDe** [comment]
@loganfsmyth How is the loading of babel.config.js different to .babelrc? require can import js and json files and the path traversal is also the same isn't it?

**Comment by nicolo-ribaudo** [comment]
You can read more about them at https://babeljs.io/docs/en/config-files

**Comment by jimmywarting** [comment]
I think babel should become more strict about requiring file extension. 
ESM demand it: https://nodejs.org/docs/latest-v17.x/api/esm.html#mandatory-file-extensions

So part of me helping others to turn on the "Compile to ESM" is a daunting process. cuz it requires to go through all the files and add the `.js` extension. If babel instead had put the foot down and say "hey, i can't find './foo' you must add the extension path. then you would be encouraging more ppl to write "correct" code.  
Supporting extension-less paths is a footgun. You are enabling bad practices for others if they ever want to at some point remove the babel compiler. cuz you are not going to transpile to cjs forever, at some point it will become esm
